File: 0387-first-unique-character-in-a-string/solution.c
int firstUniqChar(char* s) {
    int allbits = 0, duplicateBits = 0;
    for (int index = 0; index < strlen(s); index++) {
        if (allbits & (1 << s[index] - 'a'))
            duplicateBits |= (1 << s[index] - 'a');
        else
            allbits |= (1 << s[index] - 'a');
    }
    for (int index = 0;index < strlen(s); index++) {
        if (!(duplicateBits & (1 << s[index] - 'a')))
            return index;
    }
    return -1;
}

File: 0387-first-unique-character-in-a-string/solution.go
type candidate struct {
    char rune
    index int
}

func firstUniqChar(s string) int {
    queue := []candidate{}
    history := [26]int{}
    for i, c := range s {
        history[c - 'a']++
        count := history[c - 'a']
        if count == 1 {
            queue = append(queue, candidate{ char: c, index: i })
        }
        for len(queue) > 0 && history[queue[0].char - 'a'] > 1 {
            queue = queue[1:]
        }
    }
    if len(queue) == 0 {
        return -1
    }
    return queue[0].index
}
